core/src/main/java/nl/nn/adapterframework/pipes/JwtPipe.java

```diff
@@ -0,0 +1,168 @@
+import com.nimbusds.jose.JWSHeader;
+import com.nimbusds.jose.JWSSigner;
+import com.nimbusds.jose.KeyLengthException;
+import com.nimbusds.jose.crypto.MACSigner;
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.JWTClaimsSet.Builder;
+import com.nimbusds.jwt.SignedJWT;
+
+import nl.nn.adapterframework.configuration.ConfigurationException;
+import nl.nn.adapterframework.core.ParameterException;
+import nl.nn.adapterframework.core.PipeLineSession;
+import nl.nn.adapterframework.core.PipeRunException;
+import nl.nn.adapterframework.core.PipeRunResult;
+import nl.nn.adapterframework.doc.Category;
+import nl.nn.adapterframework.parameters.ParameterList;
+import nl.nn.adapterframework.parameters.ParameterValueList;
+import nl.nn.adapterframework.stream.Message;
+import nl.nn.adapterframework.util.CredentialFactory;
+
+/**
+ * Creates a JWT
+ *
+ * @author Niels Meijer
+ * @since 7.9
+ */
+@Category("Basic")
+public class JwtPipe extends FixedForwardPipe {
+	private static final String SHARED_SECRET_PARAMETER_NAME = "sharedSecret";
+
+	private JWSHeader jwtHeader;
+	private JWSSigner globalSigner;
+
+	private String sharedSecret;
+	private String sharedSecretAlias;
+	private int expirationTime = 600;
+
+	@Override
+	public void configure() throws ConfigurationException {
+		parameterNamesMustBeUnique = true;
+		super.configure();
+
+		jwtHeader = new JWSHeader.Builder(JWSAlgorithm.HS256).type(JOSEObjectType.JWT).build();
+
+		if(StringUtils.isNotEmpty(sharedSecret) || StringUtils.isNotEmpty(sharedSecretAlias)) {
+			try {
+				CredentialFactory credentialFactory = new CredentialFactory(sharedSecretAlias, null, () -> sharedSecret);
+				globalSigner = new MACSigner(credentialFactory.getPassword());
+			} catch (KeyLengthException e) {
+				throw new ConfigurationException("invalid shared key", e);
+			}
+		}
+
+		if(globalSigner == null && getParameterList().findParameter(SHARED_SECRET_PARAMETER_NAME) == null) {
+			throw new ConfigurationException("must either provide a [sharedSecret] (alias) or parameter");
+		}
+	}
+
+	@Override
+	public PipeRunResult doPipe(Message message, PipeLineSession session) throws PipeRunException {
+		final JWSSigner signer;
+		Builder claimsSetBuilder = new JWTClaimsSet.Builder();
+
+		Map<String, Object> parameterMap = getParameterValueMap(message, session);
+		if(parameterMap != null) {
+			Object sharedKey = parameterMap.remove(SHARED_SECRET_PARAMETER_NAME);
+			parameterMap.forEach(claimsSetBuilder::claim);
+
+			signer = getSigner(sharedKey);
+		} else {
+			signer = globalSigner;
+		}
+
+		if(expirationTime > 0) {
+			Date expirationDate = Date.from(Instant.now().plusSeconds(expirationTime));
+			claimsSetBuilder.expirationTime(expirationDate);
+		}
+		claimsSetBuilder.issueTime(Date.from(Instant.now()));
+
+		String jwtToken = createAndSignJwtToken(signer, claimsSetBuilder.build());
+		return new PipeRunResult(getSuccessForward(), Message.asMessage(jwtToken));
+	}
```

Comment:
Just wondering: every separate message that passes the pipe fetches a new JWT session token? Because every message is a new HTTP session?

Reply:
In this case, every time the pipe is called a new token will be generated.

Reply:
Yes indeed, that's what I figured. So, if the pipe encounters a whole flow of messages, it will cause some overhead. Perhaps add this in the documentation, that this is how it works?

Reply:
That's how every pipe/sender (with parameters) work, but I don't think we have to specify that. If it has to be dynamic, it will add overhead. The Developer won't use a dynamic option unless they have to.

Reply:
Ok, clear.